app/core/middleware.py

```diff
@@ -0,0 +1,47 @@
+from logging import getLogger
+from typing import Callable
+
+from django.conf import settings
+from django.http.request import HttpRequest
+from django.http.response import HttpResponse
+
+LOGGER = getLogger(__name__)
+
+
+def _set_can_modify(request: HttpRequest, state: bool) -> HttpRequest:
+    request.META[settings.KONG_MODIFICATION_HEADER] = state
+    return request
+
+
+def _set_authenticated(request: HttpRequest, state: bool) -> HttpRequest:
+    request.META[settings.KONG_AUTH_HEADER] = state
+    return request
+
+
+def kong_authentication(get_response: Callable[[HttpRequest], HttpResponse]) \
+        -> Callable[[HttpRequest], HttpResponse]:
+    def middleware(request: HttpRequest):
+        authenticated = False
+        can_modify = False
+
+        groups_header = request.META.get(settings.CONSUMER_GROUPS_HEADER, None)
+
+        if groups_header:
+            groups = [group.strip() for group in groups_header.split(',')]
+
+            LOGGER.debug('user groups: %s', groups)
+            if 'view-unpublished-content' in groups:
+                authenticated = True
+            else:
+                LOGGER.debug('setting request as not authenticated')
+
+            if 'edit-digests' in groups:
+                can_modify = True
+            else:
+                LOGGER.debug('setting request as user cannot modify digests')
+
+        request = _set_authenticated(_set_can_modify(request, can_modify), authenticated)
```

Comment:
The additional headers can be set from the outside and the api-gateway won't strip them, being them unknown to it; so they should be reset to empty/missing before we attempt to add them depending on `X-Consumer-Groups`.

Better yet if there is a way to store state on the request that does not use headers at all, as it prevents the attack (and the confusion) by design

Reply:
In its current implementation, they are always set with a value so if there were header values (e.g. someone set `Digests-Can-Preview` to `True` externally) present on the initial request the middleware would always overwrite them anyway.

You can just attach values to the `request` object itself (its just a python object and is not implementing slots) but it's just a little less 'standard' than using the headers.

Its not a big change to attach the state directly to the `request` object, should I refactor and do it?